File: Brainfuck.Python/brainfuck.py
import sys
from os import path
from typing import List

def Interpret(text : str) -> None:
    mem : List[byte] = [0] * 30000
    ptr : int = 0
    
    jmp : List[int] = []
    
    i : int = 0
    while i < len(text):
        c : char = text[i]
        
        if c == '>':
            ptr += 1
        elif c == '<':
            ptr -= 1
        elif c == '+':
            mem[ptr] = mem[ptr] + 1
        elif c == '-':
            mem[ptr] = mem[ptr] - 1
        elif c == ',':
            mem[ptr] = ord(sys.stdin.read(1))
        elif c == '.':
            print(chr(mem[ptr]), end="")
            sys.stdout.flush()
        elif c == '[':
            if mem[ptr] != 0:
                jmp.append(i)
            else:
                brackets : int = 1
                
                while brackets > 0:
                    i += 1
                    
                    if text[i] == '[':
                        brackets += 1
                    elif text[i] == ']':
                        brackets -= 1
        elif c == ']':
            if mem[ptr] != 0:
                i = jmp[-1]
            else:
                jmp.pop()
                
        i += 1

if not len(sys.argv) == 2:
    print("Supply a file parameter containing valid brainfuck code.")
    sys.exit()
    
file : str = sys.argv[1]

if not path.exists(file):
    print("File " + file + " does not exist!")
    print("Exiting.")
    sys.exit()

with open(file) as f:
    text : str = "".join(f.readlines())
    Interpret(text)

File: Brainfuck.Julia/brainfuck.jl
function Interpret(text)
    mem = Vector{UInt8}(undef, 30000)
    mem .= zero(UInt8)
    
    ptr = 1
    jmp = Vector{Int32}()
    
    i = 1
    while i <= length(text)
        c = text[i]
        if c == '>'
            ptr += 1
        elseif c == '<'
            ptr -= 1
        elseif c == '+'
            if mem[ptr] == 255
                mem[ptr] = 0
            else
                mem[ptr] += 1
            end
        elseif c == '-'
            if mem[ptr] == 0
                mem[ptr] = 255
            else
                mem[ptr] -= 1
            end
        elseif c == ','
            mem[ptr] = convert(UInt8, read(stdin, Char))
        elseif c == '.'
            print(convert(Char, mem[ptr]))
            flush(stdout)
        elseif c == '['
            if mem[ptr] != 0
                append!(jmp, i)
            else
                brackets = 1
                while brackets > 0
                    i += 1
                    
                    if text[i] == '['
                        brackets += 1
                    elseif text[i] == ']'
                        brackets -= 1
                    end
                end
            end
        elseif c == ']'
            if mem[ptr] != 0
                i = last(jmp)
            else
                pop!(jmp)
            end
        end
        
        i += 1
    end
end

if length(ARGS) != 1
    println("Supply a file parameter containing valid brainfuck code.")
    exit()
end

file = ARGS[1]

if !isfile(file)
    println("File " * file * " does not exist!")
    exit()
end

Interpret(read(file, String))
